lib/README.md

```diff
@@ -0,0 +1,144 @@
+# Gutenberg PHP
+
+## File structure
+
+Gutenberg adds features to WordPress Core using PHP hooks and filters. Some
+features, once considered stable and useful, are merged into Core during a Core
+release. Some features remain in the plugin forever or are removed.
+
+To make it easier for contributors to know which features need to be merged to
+Core and which features can be deleted, Gutenberg uses the following file
+structure for its PHP code:
+
+- `lib/experimental` - Experimental features that exist only in the plugin. They
+  are not ready to be merged to Core.
+- `lib/stable` - Stable features that exist only in the plugin. They could one
+  day be merged to Core, but not yet.
+- `lib/compat/wordpress-X.Y` - Stable features that are intended to be merged to
+  Core in the future X.Y release, or that were previously merged to Core in the
+  X.Y release and remain in the plugin for backwards compatibility when running
+  the plugin on older versions of WordPress.
+
+## Best practices
+
+### Prefer the `wp` prefix
+
+For features that may be merged to Core, it's best to use a `wp_` prefix for functions or a `WP_` prefix for classes. 
+
+This applies to both experimental and stable features.
+
+Using the `wp_` prefix avoids us having to rename functions and classes from `gutenberg_` to `wp_` if the feature is merged to Core.
+
+Functions that are intended solely for the plugin, e.g., plugin infrastructure, should use the `gutenberg_` prefix.
+
```

Comment:
Any thoughts on naming conventions for when we're updating a function from core?

For example, this change from https://github.com/WordPress/gutenberg/pull/38681/files?w=1#diff-089523f9c993d53f6f02926d28f7a80394d93740e7d16e47e17244547b2b8f5e. The core action is removed and the plugin action is added to replace it.

Reply:
I don't think we really have a good strategy for this kind of stuff. We have a similar nightmare with `WP_Theme_JSON` needing `Gutenberg_Theme_JSON` and `WP_Theme_JSON_5_9` 😅

Using `gutenberg_` seems fine to me, though it will get tricky when the new `gutenberg_` function makes it into WP 6.0 and we want to make more changes in 6.1 We won't be able to tell whether the `wp_` function that exists is the 5.9 version or the 6.0 version.